packages/eas-cli/src/project/publish.ts

```diff
@@ -0,0 +1,308 @@
+    };
+  });
+}
+
+export function collectBundles(distRoot: string): { [key: string]: RawAsset } {
+  const bundleRoot = path.join(distRoot, 'bundles');
+  const bundlePaths = Object.fromEntries(
+    fs.readdirSync(bundleRoot).map(name => [name.split('-')[0], path.join(bundleRoot, name)])
+  );
+
+  const bundleBuffers: { [key: string]: RawAsset } = {};
+  Platforms.forEach(platform => {
+    bundleBuffers[platform] = {
+      type: 'bundle',
+      contentType: 'application/javascript',
+      buffer: fs.readFileSync(bundlePaths[platform]),
+    };
+  });
+  return bundleBuffers;
+}
+
+export function collectAssets(inputDir: string): CollectedAssets {
+  const distRoot = getDistRoot(inputDir);
+  const assets = collectUserDefinedAssets(distRoot);
+  const bundles = collectBundles(distRoot);
+  const assetsFinal: CollectedAssets = {};
+
+  let platform: PublishPlatforms;
+  for (platform of Platforms) {
+    assetsFinal[platform] = {
+      launchAsset: bundles[platform],
+      assets,
+    };
+  }
+
+  return assetsFinal;
+}
+
+export async function filterOutAssetsThatAlreadyExistAsync(
+  uniqueAssetsWithStorageKey: (RawAsset & { storageKey: string })[]
+): Promise<(RawAsset & { storageKey: string })[]> {
+  const assetMetadata = await PublishQuery.getAssetMetadataAsync(
+    uniqueAssetsWithStorageKey.map(asset => asset.storageKey)
+  );
+  const missingAssetMetadata = assetMetadata
+    .filter((am: any) => am.status !== 'EXISTS')
+    .map((am: any) => am.storageKey);
+
+  const missingAssets = uniqueAssetsWithStorageKey.filter(asset => {
+    return missingAssetMetadata.includes(asset.storageKey);
+  });
+  return missingAssets;
+}
+
+export async function uploadAssetsAsync(assetsForUpdateInfoGroup: CollectedAssets): Promise<void> {
+  let assets: RawAsset[] = [];
+  let platform: keyof CollectedAssets;
+  for (platform in assetsForUpdateInfoGroup) {
+    assets = [
+      ...assets,
+      assetsForUpdateInfoGroup[platform]!.launchAsset,
+      ...assetsForUpdateInfoGroup[platform]!.assets,
+    ];
+  }
+
+  const uniqueAssets = uniqBy<
+    RawAsset & {
+      storageKey: string;
+    }
+  >(
+    assets.map(asset => {
+      return {
+        ...asset,
+        storageKey: getStorageKeyForAsset(asset),
+      };
+    }),
+    'storageKey'
+  );
+
+  let missingAssets = await filterOutAssetsThatAlreadyExistAsync(uniqueAssets);
+  const { specifications } = await PublishMutation.getUploadURLsAsync(
+    missingAssets.map((ma: any) => ma.contentType)
+  );
+
+  await Promise.all(
+    missingAssets.map((missingAsset, i) => {
+      const presignedPost: PresignedPost = JSON.parse(specifications[i]);
+      return uploadWithPresignedPostAsync(missingAsset.buffer, presignedPost);
+    })
+  );
+
+  // Wait up to TIMEOUT_LIMIT for assets to be uploaded and processed
+  const start = Date.now();
+  while (missingAssets.length > 0) {
+    if (process.env.NODE_ENV !== 'test') {
+      await new Promise(resolve => setTimeout(resolve, 1000));
```

Comment:
I think it might be nice to slow down the pace of polling (e.g. exponential backoff) when more time has passed. For example if it's already taken 30 seconds, we might as well just check every few seconds rather than once per second. This way the total number of requests that hit the server could be reduced, while the time to wait would still be roughly the same.

This is just a nice to have, though, feel free to not implement this suggestion.

Reply:
I added a linear back off as exponential seemed a bit aggressive.